File: clients/android/app/src/main/java/live/yunmao/app/network/YunmaoApi.kt
package live.yunmao.app.network

import io.ktor.client.HttpClient
import io.ktor.client.engine.HttpClientEngine
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.websocket.WebSockets
import io.ktor.client.request.bearerAuth
import io.ktor.client.request.get
import io.ktor.client.request.post
import io.ktor.client.request.setBody
import io.ktor.client.call.body
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.json.Json
import live.yunmao.app.model.*

// YunmaoApi：基于 Ktor 的 HTTP 客户端，统一注入 Bearer JWT。
// 与 iOS YunmaoAPI / Web YunmaoApi 路径一致。
class YunmaoApi(
    private val baseUrl: String,
    private val tokenProvider: () -> String?,
    engine: HttpClientEngine = OkHttp.create(),
) {
    private val client = HttpClient(engine) {
        install(ContentNegotiation) {
            json(Json {
                ignoreUnknownKeys = true
                explicitNulls = false
                encodeDefaults = true
            })
        }
        install(WebSockets)
    }

    private fun io.ktor.client.request.HttpRequestBuilder.applyAuth() {
        tokenProvider()?.let { bearerAuth(it) }
    }

    private suspend inline fun <reified T> get(path: String): T =
        client.get("$baseUrl$path") {
            tokenProvider()?.let { bearerAuth(it) }
        }.body()

    private suspend inline fun <reified Req, reified Resp> postJson(path: String, body: Req): Resp =
        client.post("$baseUrl$path") {
            tokenProvider()?.let { bearerAuth(it) }
            contentType(ContentType.Application.Json)
            setBody(body)
        }.body()

    suspend fun login(phone: String, code: String): AuthToken =
        postJson("/api/v1/auth/login", mapOf("phone" to phone, "code" to code))

    suspend fun listRooms(): List<Room> = get<RoomList>("/api/v1/rooms").rooms

    suspend fun subscription(roomId: String): RoomSubscription =
        get("/api/v1/rooms/$roomId/subscription")

    suspend fun feed(req: FeedRequest): FeedResponse = postJson("/api/v1/feed", req)

    suspend fun wallet(userId: String): Wallet = get("/api/v1/wallets/$userId")

    suspend fun createPrepay(orderId: String, channel: String, amountFen: Long): PrepayResponse =
        client.post("$baseUrl/api/v1/orders/$orderId/prepay?channel=$channel") {
            tokenProvider()?.let { bearerAuth(it) }
            contentType(ContentType.Application.Json)
            setBody(mapOf("amount_fen" to amountFen, "subject" to "yunmao 充值"))
        }.body()

    fun close() = client.close()
}

File: clients/ios/YunmaoApp/Sources/YunmaoApp/Network/YunmaoAPI.swift
// YunmaoAPI：URLSession + async/await 客户端，统一注入 JWT。
//
// 端点（与后端 chi 路由对齐）：
//   - POST  /api/v1/auth/login          → AuthToken
//   - GET   /api/v1/rooms               → [Room]
//   - GET   /api/v1/rooms/{id}/subscription → RoomSubscription
//   - GET   /v1/rooms/{id}/ice-servers  → IceServersResponse
//   - POST  /api/v1/feed                → FeedResponse
//   - GET   /api/v1/wallets/{user_id}   → Wallet
//   - POST  /api/v1/orders/{id}/prepay  → PrepayResponse
//   - POST  /api/v1/pay/webhook/appleiap → ack
//
// 错误：HTTPError（非 2xx）+ DecodingError + URLError。

import Foundation

public enum YunmaoAPIError: Error, LocalizedError {
    case http(Int, String)
    case noToken
    case invalidResponse

    public var errorDescription: String? {
        switch self {
        case .http(let code, let body): return "HTTP \(code): \(body)"
        case .noToken: return "missing auth token"
        case .invalidResponse: return "invalid response"
        }
    }
}

public actor YunmaoAPI {
    public let baseURL: URL
    private let session: URLSession
    private var tokenProvider: () -> String?

    public init(baseURL: URL, session: URLSession = .shared, tokenProvider: @escaping () -> String?) {
        self.baseURL = baseURL
        self.session = session
        self.tokenProvider = tokenProvider
    }

    public func updateTokenProvider(_ provider: @escaping () -> String?) {
        self.tokenProvider = provider
    }

    // MARK: - generic

    public func get<T: Decodable>(_ path: String, query: [URLQueryItem] = [], as type: T.Type = T.self) async throws -> T {
        let req = try request(path: path, method: "GET", query: query, body: nil)
        return try await execute(req)
    }

    public func post<B: Encodable, T: Decodable>(_ path: String, body: B, as type: T.Type = T.self) async throws -> T {
        let data = try JSONEncoder.yunmao.encode(body)
        var req = try request(path: path, method: "POST", body: data)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await execute(req)
    }

    public func postRaw(_ path: String, body: Data, contentType: String = "application/json") async throws -> Data {
        var req = try request(path: path, method: "POST", body: body)
        req.setValue(contentType, forHTTPHeaderField: "Content-Type")
        let (data, resp) = try await session.data(for: req)
        try assertOK(resp, body: data)
        return data
    }

    // MARK: - typed endpoints

    public func login(phone: String, code: String) async throws -> AuthToken {
        struct In: Codable { let phone: String; let code: String }
        return try await post("/api/v1/auth/login", body: In(phone: phone, code: code))
    }

    public func listRooms() async throws -> [Room] {
        struct Wrap: Codable { let rooms: [Room] }
        let w: Wrap = try await get("/api/v1/rooms")
        return w.rooms
    }

    public func subscription(roomID: String) async throws -> RoomSubscription {
        try await get("/api/v1/rooms/\(roomID)/subscription")
    }

    public func iceServers(roomID: String) async throws -> IceServersResponse {
        try await get("/v1/rooms/\(roomID)/ice-servers")
    }

    public func feed(_ req: FeedRequest) async throws -> FeedResponse {
        try await post("/api/v1/feed", body: req)
    }

    public func wallet(userID: String) async throws -> Wallet {
        try await get("/api/v1/wallets/\(userID)")
    }

    public func appleIAPWebhook(payload: Data) async throws {
        _ = try await postRaw("/api/v1/pay/webhook/appleiap", body: payload)
    }

    // MARK: - core

    private func request(path: String, method: String, query: [URLQueryItem] = [], body: Data?) throws -> URLRequest {
        var components = URLComponents(url: baseURL.appendingPathComponent(path), resolvingAgainstBaseURL: false)!
        if !query.isEmpty { components.queryItems = query }
        guard let url = components.url else { throw YunmaoAPIError.invalidResponse }
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.httpBody = body
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token = tokenProvider() {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return req
    }

    private func execute<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp) = try await session.data(for: req)
        try assertOK(resp, body: data)
        return try JSONDecoder.yunmao.decode(T.self, from: data)
    }

    private func assertOK(_ resp: URLResponse, body: Data) throws {
        guard let http = resp as? HTTPURLResponse else { throw YunmaoAPIError.invalidResponse }
        if !(200..<300).contains(http.statusCode) {
            throw YunmaoAPIError.http(http.statusCode, String(data: body, encoding: .utf8) ?? "")
        }
    }
}

extension JSONEncoder {
    public static let yunmao: JSONEncoder = {
        let e = JSONEncoder()
        e.keyEncodingStrategy = .convertToSnakeCase
        return e
    }()
}

extension JSONDecoder {
    public static let yunmao: JSONDecoder = {
        let d = JSONDecoder()
        d.keyDecodingStrategy = .convertFromSnakeCase
        return d
    }()
}
